src/main/kotlin/app/CodeLongevity.kt

```diff
@@ -0,0 +1,266 @@
+                for (edit in editList) {
+                    val delStart = edit.getBeginA()
+                    val delEnd = edit.getEndA()
+                    val delCount = edit.getLengthA()
+                    var insStart = edit.getBeginB()
+                    var insEnd = edit.getEndB()
+                    val insCount = edit.getLengthB()
+                    println("del ($delStart, $delEnd), ins ($insStart, $insEnd)");
+
+                    // Deletion case. Chase down the deleted lines through the
+                    // history.
+                    if (delCount > 0) {
+                        var tmpLines = ArrayList<RevCommitLine>(delCount)
+                        var idx = delStart;
+                        while (idx < delEnd) {
+                            tmpLines.add(RevCommitLine(commit, oldPath, idx));
+                            idx++
+                        }
+
+                        lines.addAll(delStart, tmpLines)
+                    }
+
+                    // Insertion case. Report it.
+                    if (insCount > 0) {
+                        val fileLoader = repo.open(newId)
+                        val fileText = RawText(fileLoader.getBytes())
+
+                        for (idx in insStart .. insEnd - 1) {
+                            val from = RevCommitLine(commit, newPath, idx)
+                            var to = lines.get(idx)
+                            val cl = CodeLine(from, to, fileText.getString(idx))
+                            codeLines.add(cl)
+                        }
+                        lines.subList(insStart, insEnd).clear()
+                    }
+                }
+
+                // File was renamed, tweak the files map.
+                if (diff.changeType == DiffEntry.ChangeType.RENAME) {
+                    files.set(oldPath, files.remove(newPath)!!)
+                }
+            }
+            commit = parentCommit
+        }
+        while (parentCommit != null && parentCommit != tail)
+
+        // If a tail revision was given then the map has to contain unclaimed
+        // code lines, i.e. the lines added before the tail revision. Push
+        // them all into the result lines list, so the caller can update their
+        // ages properly.
+        if (tail != null) {
+            for ((file, lines) in files) {
+                for (idx in 0 .. lines.size - 1) {
+                    val from = RevCommitLine(tail, file, idx)
+                    val cl = CodeLine(from, lines[idx], "no data (too lazy to compute)")
+                    codeLines.add(cl)
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns lines array of a text file of the given revision.
+     */
+    private fun getFileLines(fileId: ObjectId, filePath: String,
+                             commit: RevCommit) : ArrayList<RevCommitLine>? {
+        val fileLoader = repo.open(fileId)
+        if (RawText.isBinary(fileLoader.openStream())) {
+            return null
```

Comment:
If I'm told twice that I'm wrong that I has to be wrong, let me defend myself though :). 

I'm not sure how I break (ii) since Kotlin takes care of it on a compiling stage, and not sure why we have to have an exception in the first place (i). I think the concept of exceptions is not quite suitable for this case: exceptions is a good way to say something wrong happens, mostly to react to unexpected (bad) input. If you need to catch a problem in your own code, then assertions concept is used. The assertions are seen in debug build only and totally ignored in release builds, and thus they have zero impact (cost) for release builds. Assertions unlike exceptions are used for something that should *never* happen, to debug your app. I believe that this case suits better for assertions than exceptions, regardless this concept is present in Kotlin or not.

So getFileLines is called twice: a) when the files map is created and b) when deleted files are processed. I'd say assertions is more appropriate for the latter case, and no exceptions is for the first case, otherwise we end up with try/catch constructions, which is somewhat ugly, because binary files is a normal thing for a common repo. You don't suggest to use exceptions mechanism as a return value, like hey here's is a binary file, and thus no lines for you, right?

Anyway, let me unwrap body of this function right into code, those few lines are not worth such long debates and probably don't even deserve its own function, at least it will be less ambiguous and not so concernful.